### PR TITLE
Fix bugs found by unit tests

### PR DESCRIPTION
**Pull request by mjuric**
Should fix the remaining issues @migueldvb reported in issue #55

### CODE REVIEW

**Comment by migueldvb** [comment]
Thank you @mjuric for fixing this problem. All the tests pass using python 2.7.11 on Linux. However I am getting a `RuntimeError` when I run the tests with python 3.5 (`python tests/testAll.py`):

```
Traceback (most recent call last):
  File "tests/testAll.py", line 6, in <module>
    import testCommon
  File "/mnt/home/miguel/LSST/eups/tests/testCommon.py", line 8, in <module>
    for var in setupvars:
  File "/usr/lib/python3.5/_collections_abc.py", line 653, in __iter__
    yield from self._mapping
  File "/usr/lib/python3.5/os.py", line 702, in __iter__
    for key in self._data:
RuntimeError: dictionary changed size during iteration
```
I will look into this issue. I think it is fine to merge this PR as it solves the problem with the failing tests in python 2.7.

**Comment by r-owen** [comment]
I hope you will be able to figure out the runtime error before merging. It seems likely that a dict is being manipulated while iterating over it, even in python 2.7, which could cause unpredictable behavior.

**Comment by timj** [comment]
When I run the tests with python3.4 I get different errors (after I've manually fixed the `next()` iterator -- I don't quite understand how it's working for @migueldvb on python 3.5):
```
$ python tests/testAll.py 
.............................................................................................................FF
======================================================================
FAIL: testDyldLibraryPathRetention_csh (testDyldLibraryPath.DyldLibraryPath)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/timj/work/eups/tests/testDyldLibraryPath.py", line 75, in testDyldLibraryPathRetention_csh
    self._run_script("/bin/csh", csh_test % 'a/b/c', "a/b/c\n/foo/bar:a/b/c\na/b/c\n")
  File "/Users/timj/work/eups/tests/testDyldLibraryPath.py", line 67, in _run_script
    self.assertEqual(output, expect)
AssertionError: 'a/b/c\nsetup: must use keyword argument for key function\nsetup: Command not found.\na/b/c\nunsetup: Command not found.\na/b/c\n' != 'a/b/c\n/foo/bar:a/b/c\na/b/c\n'

======================================================================
FAIL: testDyldLibraryPathRetention_sh (testDyldLibraryPath.DyldLibraryPath)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/timj/work/eups/tests/testDyldLibraryPath.py", line 70, in testDyldLibraryPathRetention_sh
    self._run_script("/bin/sh", sh_test % 'a/b/c', "a/b/c\n/foo/bar:a/b/c\na/b/c\n")
  File "/Users/timj/work/eups/tests/testDyldLibraryPath.py", line 67, in _run_script
    self.assertEqual(output, expect)
AssertionError: "a/b/c\nsetup: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte\n/bin/sh: line 5: setup: command not found\na/b/c\n/bin/sh: line 8: unsetup: command not found\na/b/c\n" != 'a/b/c\n/foo/bar:a/b/c\na/b/c\n'

----------------------------------------------------------------------
Ran 111 tests in 3.087s

FAILED (failures=2)
```

**Comment by mjuric** [comment]
@timj @migueldvb I'm confused -- I just tried running with Python 3.5, and I got a slew of errors throughout the code (e.g., `sort` method doesn't take a `cmp` any more, then there were a number of `xrange` errors, etc.). To give you a taste:
```
[mjuric@gallifrey (bugfixes-found-by-unittests) ~/projects/eups]$ (cd tests/; source ../bin/setups.sh; python testAll.py)
EEE.E....E....EE...E.E.EE..FE...E.......EE..EEEEEEEEEEE...E.......EE.......E.EEEE......E.EE.EEEEEEE..........FF
```
I don't see how you're getting only two failures in tests -- I must be doing something really dumb?

**Comment by mjuric** [comment]
OK, I made some Python 3 compatibility fixes with on 2df3eb8 with the delicate touch and sophistication of a man wielding a machete and in a hurry to traverse a jungle. 

The unit tests now seem to pass both on 2.7 and 3.5, though I'm still confused as to how you got it to work at all w/o this (and am looking forward to being shown tomorrow that I forgot to set a simple switch, or something like that :-) ).

**Comment by migueldvb** [comment]
@mjuric I still get some import errors when I run the tests on python 3.5 with  2df3eb8. I think that I am doing something wrong.  How did you configure and install eups?

**Comment by migueldvb** [comment]
Now it is working well for me using a different root directory for each version of the installation. All the tests pass in python 2.7.11 and 3.5.1. This is how I installed things in python 3:
```
./configure --with-python=/usr/bin/python3 --with-eups=~/eups/products/
make install
```
Thank you for taking care of all the compatibility problems.

File: python/eups/Eups.py [hunk 1]
```diff
@@ -1367,7 +1368,7 @@ def _findProductsByExpr(self, name, expr, eupsPathDirs, flavor, noCache):
                 if len(products) == 0: 
                     continue
 
-                products = filter(lambda z: self.version_match(z.version, expr), products)
+                products = list(filter(lambda z: self.version_match(z.version, expr), products))
```
Comment (timj) [inline python/eups/Eups.py]: I think all these lambdas should be replaced with code like:
```python
products = [z for z in products if self.version_match(z.version, expo)]
```

File: python/eups/Eups.py [hunk 2]
```diff
@@ -1377,7 +1378,7 @@ def _findProductsByExpr(self, name, expr, eupsPathDirs, flavor, noCache):
                 # consult the cache
                 try: 
                     vers = self.versions[root].getVersions(name, flavor)
-                    vers = filter(lambda z: self.version_match(z, expr), vers)
+                    vers = list(filter(lambda z: self.version_match(z, expr), vers))
```
Comment (timj) [inline python/eups/Eups.py]:
```python
vers = [z for z in vers if self.version_match(z, expr)]
```

File: python/eups/Eups.py [hunk 3]
```diff
@@ -1405,8 +1406,8 @@ def _selectPreferredProduct(self, products, preferredTags=None):
             tag = self.tags.getTag(tag)  # should not fail
             if tag.name == "latest":
                 # find the latest version; first order the versions
-                vers = map(lambda p: p.version, products)
-                vers.sort(self.version_cmp)
+                vers = list(map(lambda p: p.version, products))
```
Comment (timj) [inline python/eups/Eups.py]:
```python
vers = [p.version for p in products]
```

File: python/eups/Eups.py [hunk 4]
```diff
@@ -1679,7 +1680,7 @@ def version_match(self, vname, expr):
         """Return vname if it matches the logical expression expr"""
 
         expr0 = expr
-        expr = filter(lambda x: not re.search(r"^\s*$", x), re.split(r"\s*(%s|\|\||\s)\s*" % self._relop_re.pattern, expr0))
+        expr = list(filter(lambda x: not re.search(r"^\s*$", x), re.split(r"\s*(%s|\|\||\s)\s*" % self._relop_re.pattern, expr0)))
```
Comment (timj) [inline python/eups/Eups.py]:
```python
expr = [x for x in re.split(r"\s*(%s|\|\||\s)\s*" % self._relop_re.pattern, expr0) if not re.search(r"^\s*$", x)]
```

File: python/eups/Eups.py [hunk 5]
```diff
@@ -3278,8 +3279,8 @@ def _remove(self, productName, versionName, recursive, checkRecursive, topProduc
         productsToRemove = []
         for product, o, recursionDepth in deps:
             if checkRecursive:
-                usedBy = filter(lambda el: el[0] != topProduct or el[1] != topVersion,
-                                userInfo.users(product.name, product.version))
+                usedBy = list(filter(lambda el: el[0] != topProduct or el[1] != topVersion,
+                                userInfo.users(product.name, product.version)))
```
Comment (timj) [inline python/eups/Eups.py]:
```python
usedBy = [el for el in userInfo.users(product.name, product.version) if el[0] != topProduct or el[1] != topVersion]
```

File: python/eups/Eups.py [hunk 6]
```diff
@@ -3659,7 +3660,7 @@ def selectVRO(self, tag=None, productDir=None, versionName=None, dbz=None, inexa
             if self.exact_version:          # this is a property of self
                 self.makeVroExact()
             if inexact_version:             # this is a request to not process type:exact in the vro
-                self._vro = filter(lambda el: el != "type:exact", self._vro)
+                self._vro = list(filter(lambda el: el != "type:exact", self._vro))
```
Comment (timj) [inline python/eups/Eups.py]:
```python
self._vro = [el for el in self._vro if el != "type:exact"]
```

File: python/eups/Eups.py [hunk 7]
```diff
@@ -3737,7 +3738,7 @@ def makeVroExact(self):
 
         if tagVroEntries:
             if movedTags:
-                if not filter(lambda x: re.search(r"^warn:[01]", x), vro):
+                if not list(filter(lambda x: re.search(r"^warn:[01]", x), vro)):
```
Comment (timj) [inline python/eups/Eups.py]:
```python
if not [x for x in vro if re.search(r"^warn:[01]", x)]:
```

File: python/eups/VersionParser.py [hunk 8]
```diff
@@ -29,7 +29,7 @@ class VersionParser(object):
     def __init__(self, exprStr):
         exprStr = re.sub(r"['\"]([^'\"]+)['\"]", r"\1", exprStr)
         self._tokens = re.split(r"(\$\??{[^}]+}|[\w.+]+|\s+|==|!=|<=|>=|[()<>])", exprStr)
-        self._tokens = filter(lambda p: p and not re.search(r"^\s*$", p), self._tokens)
+        self._tokens = list(filter(lambda p: p and not re.search(r"^\s*$", p), self._tokens))
```
Comment (timj) [inline python/eups/VersionParser.py]:
```python
self._tokens = [p for p in self._tokens if p and not re.search(r"^\s*$", p)]
```

File: python/eups/db/Database.py [hunk 9]
```diff
@@ -691,7 +692,7 @@ def isNewerThan(self, timestamp, dbrootdir=None):
 
         if not dbrootdir:
             dbrootdir = self.dbpath
-        proddirs = map(lambda d: os.path.join(self.dbpath, d), self.findProductNames())
+        proddirs = list(map(lambda d: os.path.join(self.dbpath, d), self.findProductNames()))
```
Comment (timj) [inline python/eups/db/Database.py]:
```python
proddirs = [os.path.join(self.dbpath, d) for d in self.findProductNames()]
```

File: python/eups/db/VersionFile.py [hunk 10]
```diff
@@ -199,7 +199,7 @@ def makeProducts(self):
         return Product instances for all of the flavors declared in the file.
         @return Product[] :
         """
-        return map(lambda x: self.makeProduct(x), self.info.keys())
+        return list(map(lambda x: self.makeProduct(x), self.info.keys()))
```
Comment (timj) [inline python/eups/db/VersionFile.py]:
```python
return [self.makeProduct(x) for x in list(self.info.keys())]
```

File: python/eups/distrib/Repositories.py [hunk 11]
```diff
@@ -60,7 +60,7 @@ def __init__(self, pkgroots, options=None, eupsenv=None,
                             sys.stderr)
         """
         if isinstance(pkgroots, str):
-            pkgroots = map(lambda p: p.strip(), pkgroots.split("|"))
+            pkgroots = list(map(lambda p: p.strip(), pkgroots.split("|")))
```
Comment (timj) [inline python/eups/distrib/Repositories.py]:
```python
pkgroots = [p.strip() for p in pkgroots.split("|")]
```

File: python/eups/distrib/Repositories.py [hunk 12]
```diff
@@ -948,7 +948,7 @@ def clean(self, product, version, flavor=None, options=None,
                         self.distribClean(product,version,pkgroot,distId,flavor)
                 
                 # make sure this directory is not declared for any product
-                installDirs = map(lambda x: x.dir, self.eups.findProducts())
+                installDirs = list(map(lambda x: x.dir, self.eups.findProducts()))
```
Comment (timj) [inline python/eups/distrib/Repositories.py]:
```python
installDirs = [x.dir for x in self.eups.findProducts()]
```

File: python/eups/distrib/Repository.py [hunk 13]
```diff
@@ -144,19 +144,19 @@ def _getPackageLookup(self):
...
 
         # now sort the contents
-        keys = lookup.keys()
+        keys = list(lookup.keys())
         keys.sort()
         lookup["_sortOrder"] = keys
 
         for prod in lookup["_sortOrder"]:
-            keys = filter(lambda f: f != "generic", lookup[prod].keys())
+            keys = list(filter(lambda f: f != "generic", lookup[prod].keys()))
```
Comment (timj) [inline python/eups/distrib/Repository.py]:
```python
keys = [f for f in list(lookup[prod].keys()) if f != "generic"]
```

File: python/eups/distrib/Repository.py [hunk 14]
```diff
@@ -337,16 +337,16 @@ def _listLatestProducts(self, product, flavor):
         for p in prods:
             names[p[0]] = 1
             flavors[p[2]] = 1
-        names = names.keys()
+        names = list(names.keys())
         names.sort()
-        flavors = filter(lambda f: f != "generic", flavors.keys())
+        flavors = list(filter(lambda f: f != "generic", flavors.keys()))
```
Comment (timj) [inline python/eups/distrib/Repository.py]:
```python
flavors = [f for f in list(flavors.keys()) if f != "generic"]
```

File: python/eups/distrib/Repository.py [hunk 15]
```diff
@@ -337,16 +337,16 @@ def _listLatestProducts(self, product, flavor):
...
         flavors.sort()
         flavors.insert(0, "generic")
         
         for name in names:
             for flav in flavors:
-                latest = filter(lambda p: p[0] == name and p[2] == flav, prods)
-                latest.sort(lambda a,b: self.eups.version_cmp(a[1],b[1]))
+                latest = list(filter(lambda p: p[0] == name and p[2] == flav, prods))
```
Comment (timj) [inline python/eups/distrib/Repository.py]:
```python
latest = [p for p in prods if p[0] == name and p[2] == flav]
```

File: python/eups/distrib/server.py [hunk 16]
```diff
@@ -116,9 +116,9 @@ def getTagNames(self, flavor=None, noaction=False):
         are available on the server, where * is a tag name.  The flavor 
         parameter is ignored.
         """
-        return map(lambda x: x[:-5], 
+        return list(map(lambda x: x[:-5], 
                    filter(lambda f: f.endswith(".list"), 
-                          self.listFiles("", noaction)))
+                          self.listFiles("", noaction))))
```
Comment (timj) [inline python/eups/distrib/server.py]:
```python
return [x[:-5] for x in [f for f in self.listFiles("", noaction) if f.endswith(".list")]]
```

File: python/eups/distrib/server.py [hunk 17]
```diff
@@ -1198,8 +1198,8 @@ def listDir(self, noaction=False):
             return []
         else:
             if os.path.isdir(self.loc):
-                return filter(lambda f: os.path.isfile(os.path.join(self.loc,f)), 
-                              os.listdir(self.loc))
+                return list(filter(lambda f: os.path.isfile(os.path.join(self.loc,f)), 
+                              os.listdir(self.loc)))
```
Comment (timj) [inline python/eups/distrib/server.py]:
```python
return [f for f in os.listdir(self.loc) if os.path.isfile(os.path.join(self.loc,f))]
```

**Comment by timj** [comment]
I started commenting on the `list(map(filter...)` but at this point I wonder if I should be applying these on the branch rather than as comments (and `futurize` is good at these sorts of things).

**Comment by r-owen** [comment]
A branch would be much appreciated. I've been getting an email for each one you find, which is not useful.

Replacing try/finally blocks that are intended to close files with "with open(...) as f:" is another useful modernization if anyone has time. I converted one file, then realized there was a lot more to do, so I stopped.

**Comment by mjuric** [comment]
@timj Yes, having these on a branch would be great. I won't be working on EUPS today so it's all yours.

@r-owen Yes, but I think that's a Python 2.7 feature and we still have to be 2.6 compatible.

**Comment by timj** [comment]
@mjuric who said we had to be 2.6 compatible? :smile: Wasn't @RobertLuptonTheGood okay with 2.7 and above?

Do you want me to fix up this branch with the list comprehensions?

**Comment by r-owen** [comment]
The with statement was added in python 2.5!

**Comment by timj** [comment]
I've pushed some cleanups in 01cf628.

**Comment by mjuric** [comment]
@timj Merci beaucoup! Hope to get to it later this evening.

Re Python 2.7 -- I though the discussion on #52 converged to 2.6 compatibility. That's still the default Python in CentOS 6 (and there are many CentOS 6 deployments out there). I don't think we lose much (the most annoying things missing are set literals, `OrderedDicts`, and with statements with multiple context managers (which is what I was mis-remembering above -- I thought the whole `as foo` part was added in 2.7, which would've been incredibly annoying to lose)).

**Comment by timj** [comment]
Much to my surprise, all tests pass on Python 3 for me now. I think that probably just means that our test suite has lots of holes in it...

**Comment by migueldvb** [comment]
There are a few uses of `map` and `filter` in testDb and testCommon that can be replaced by list comprehensions as well. Should I open a pull request on this branch with those changes?

**Comment by timj** [comment]
@migueldvb I'll fix those since I've got the editor open. I'll include them in this PR.

**Comment by migueldvb** [comment]
@timj OK thanks

**Comment by timj** [comment]
Pushed. I rewrote one of the map/filter/maps as a for because it was relying on side effects rather than returning a list.

**Comment by timj** [comment]
Should add that all tests are still passing on 2.7 and 3.4 for me and I can use python 3.4 to `eups list` my normal stack.

**Comment by migueldvb** [comment]
All tests pass for me on python 3.5.

**Comment by timj** [comment]
I just rebased this on current master and force pushed so it's ready for someone to push the Big Green Button.

**Comment by migueldvb** [comment]
@timj in line 601 of `testsDb.py` you could replace 
```
prod = [d for d in prods if d.version == "1.5.7.1"][0]
```
by 
```
prod = next(d for d in prods if d.version == "1.5.7.1")
```

**Comment by timj** [comment]
Done in commit daa1342

**Comment by mjuric** [comment]
@timj @migueldvb Thanks -- I'll push the green button later tonight; I want to clean up (merge/split/reword commit messages) the patchset first (so when we do start seeing regressions, they're easier to trace).

PS: Btw., if any of you get a chance, could you test if `eups distrib install` still works? I think `server.py` got some changes, but I'm not sure to what degree it's exercised by the tests.

**Comment by mjuric** [comment]
@timj @migueldvb Merged!

A few final changes I made, for our joint future reference:
* I avoided materializing a list where not necessary (see a911261). I think that's probably a good practice to follow
* I replaced a number of list comprehensions with more readable equivalents, as they were even harder to understand than their functional equivalents (see c1d122c).

All tests are passing on 2.7 and 3.5. As an interesting (though potentially not significant) data point, it takes 3.7 seconds to execute them on 2.7, and 4.35 seconds on 3.5.

**Pull request merged**